packages/game-passes/contracts/GamePasses.sol

```diff
@@ -0,0 +1,1161 @@
+     * @dev Used for user-initiated token transformations
+     * @dev Contract must not be paused
+     * @dev Reverts if:
+     *      - Contract is paused
+     *      - from address doesn't match msg.sender
+     *      - Burn token is not configured
+     *      - Mint token is not configured
+     *      - Signature is invalid or expired
+     *      - Max supply would be exceeded for mint token
+     *      - Burn operation fails (insufficient balance)
+     */
+    function burnAndMint(
+        address from,
+        uint256 burnId,
+        uint256 burnAmount,
+        uint256 mintId,
+        uint256 mintAmount,
+        uint256 deadline,
+        bytes memory signature
+    ) external whenNotPaused {
+        address caller = _msgSender();
+
+        if (from != caller) {
+            revert InvalidSender(from, caller);
+        }
+
+        TokenConfig storage burnConfig = tokenConfigs[burnId];
+        if (!burnConfig.isConfigured) {
+            revert BurnMintNotConfigured(burnId);
+        }
+
+        // Check if mint token is configured and respects max supply
+        TokenConfig storage mintConfig = tokenConfigs[mintId];
+        if (!mintConfig.isConfigured) {
+            revert TokenNotConfigured(mintId);
+        }
+
+        BurnAndMintRequest memory request = BurnAndMintRequest({
+            caller: caller,
+            burnId: burnId,
+            burnAmount: burnAmount,
+            mintId: mintId,
+            mintAmount: mintAmount,
+            deadline: deadline,
+            nonce: nonces[caller]++
+        });
+
+        verifyBurnAndMintSignature(request, signature);
+
+        _checkMaxSupply(mintId, mintAmount);
+
+        // Burn first
+        _burn(caller, burnId, burnAmount);
+
+        // Then mint new token
+        _mint(caller, mintId, mintAmount, "");
+    }
+
+    /**
+     * @notice Add or remove addresses from token transfer whitelist
+     * @param tokenId The token ID to update whitelist for
+     * @param accounts Array of addresses to update
+     * @param allowed Whether the addresses should be allowed to transfer
+     * @dev Only callable by addresses with ADMIN_ROLE
+     * @dev Token must be already configured
+     * @dev Whitelisted addresses can transfer tokens even if token is non-transferable
+     * @dev Reverts if:
+     *      - Caller doesn't have ADMIN_ROLE
+     *      - Token is not configured
+     */
+    function updateTransferWhitelist(
+        uint256 tokenId,
+        address[] calldata accounts,
+        bool allowed
+    ) external onlyRole(ADMIN_ROLE) {
+        TokenConfig storage config = tokenConfigs[tokenId];
+
+        if (!config.isConfigured) {
+            revert TokenNotConfigured(tokenId);
+        }
+
+        for (uint256 i = 0; i < accounts.length; i++) {
+            config.transferWhitelist[accounts[i]] = allowed;
+            emit TransferWhitelistUpdated(tokenId, accounts[i], allowed);
```

Comment:
as a suggestion: if we have a list of addresses in the admin role maybe it can help to add the _msgSender (`operator`)  to the event.

Reply:
Thats a good point, we can emit the callers wallet address.
Added: https://github.com/thesandboxgame/sandbox-smart-contracts/pull/1619/commits/dcfe8b87e7dc583bd9bc147374e8de78f87f4177